semtypes/src/main/java/io/ballerina/types/subtypedata/XmlSubtype.java

```diff
@@ -0,0 +1,137 @@
+            SubtypeData ro = Core.getComplexSubtypeData(cct, UniformTypeCode.UT_XML_RO);
+            ro = (ro instanceof AllOrNothingSubtype) ? ro : makeSequence(true, (XmlSubtype) ro);
+
+            SubtypeData rw = Core.getComplexSubtypeData(cct, UniformTypeCode.UT_XML_RW);
+            rw = (rw instanceof AllOrNothingSubtype) ? rw : makeSequence(false, (XmlSubtype) rw);
+
+            return createXmlSemtype(ro, rw);
+        }
+    }
+
+    private static SubtypeData makeSequence(boolean roPart, XmlSubtype d) {
+        int primitives = XmlPrimitive.XML_PRIMITIVE_NEVER | d.primitives;
+        int atom = d.primitives &
+                (roPart ? XmlPrimitive.XML_PRIMITIVE_RO_SINGLETON : XmlPrimitive.XML_PRIMITIVE_SINGLETON);
+        Bdd sequence = BddCommonOps.bddUnion(BddCommonOps.bddAtom(RecAtom.createRecAtom(atom)), d.sequence);
+        return createXmlSubtype(roPart, primitives, sequence);
+    }
+
+    public static ComplexSemType createXmlSemtype(SubtypeData ro, SubtypeData rw) {
+        if (ro instanceof AllOrNothingSubtype && ((AllOrNothingSubtype) ro).isNothingSubtype()) {
+            return ComplexSemType.createComplexSemType(0, UniformSubtype.from(UniformTypeCode.UT_XML_RW, rw));
+        } else if (rw instanceof AllOrNothingSubtype && ((AllOrNothingSubtype) rw).isNothingSubtype()) {
+            return ComplexSemType.createComplexSemType(0, UniformSubtype.from(UniformTypeCode.UT_XML_RO, ro));
+        } else {
+            return ComplexSemType.createComplexSemType(0, UniformSubtype.from(UniformTypeCode.UT_XML_RO, ro),
+                    UniformSubtype.from(UniformTypeCode.UT_XML_RW, rw));
+        }
+    }
+
+    public static SubtypeData createXmlSubtype(boolean isRo, int primitives, Bdd sequence) {
+        int mask = isRo ? XmlPrimitive.XML_PRIMITIVE_RO_MASK : XmlPrimitive.XML_PRIMITIVE_RW_MASK;
+        int p = primitives & mask;
+        if (sequence instanceof BddAllOrNothing && ((BddAllOrNothing) sequence).isAll() && p == mask) {
+            return AllOrNothingSubtype.createAll();
+        }
+        return createXmlSubtypeOrEmpty(p, sequence);
+    }
+
+    public static SubtypeData createXmlSubtypeOrEmpty(int primitives, Bdd sequence) {
+        if (sequence instanceof BddAllOrNothing  && ((BddAllOrNothing) sequence).isNothing() && primitives == 0) {
+            return AllOrNothingSubtype.createNothing();
+        }
+        return from(primitives, sequence);
+    }
+
+    public static int collectAllBits(Conjunction con) {
+        int allBits = 0;
+        Conjunction current = con;
+        while (current != null) {
+            allBits |= current.atom.getIndex();
+            current = current.next;
+        }
+        return allBits;
+    }
+
+    public static boolean hasTotalNegative(int allBits, Conjunction con) {
+        if (allBits == 0) {
+            return true;
+        }
+
+        Conjunction n = con;
+        while (n != null) {
+            if ((allBits & ~n.atom.getIndex()) == 0) {
```

Comment:
Instead of adding a method to `Atom` shall we create a helper method that just cast and get the index from `RecAtom`